addons/netfox/properties/history-buffer.gd.gd

```diff
@@ -0,0 +1,66 @@
+class_name _HistoryBuffer
+
+# Typed as Dictionary[int, PropertyStoreSnapshot]
+var _history: Dictionary
+
+func get_closest_tick(tick: int) -> int:
+	if _history.is_empty():
+		return -1
+	
+	if _history.has(tick):
+		return tick
+	
+	var earliest_tick: int = _history.keys().min()
+	if tick < earliest_tick:
+		return earliest_tick
+	
+	var latest_tick: int = _history.keys().max()
+	if tick > latest_tick:
+		return latest_tick
+	
+	return _history.keys() \
+		.filter(func (key): return key < tick) \
+		.max()
+
+func get_closest_snapshot(tick: int) -> PropertyStoreSnapshot:
+	var closest_tick = get_closest_tick(tick)
+	if closest_tick == -1:
+		return PropertyStoreSnapshot.new()
+	
+	return _history[closest_tick]
+
+func get_earliest_tick() -> int:
+	return _history.keys().min()
+
+func get_latest_tick() -> int:
+	return _history.keys().max()
```

Comment:
It might make sense to cache these, if these methods keep being used in the final PR. Needs benchmarking.